L1Trigger/TrackerTFP/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,207 @@
+    if (iConfig_.getParameter<bool>("CheckHistory"))
+      setup_->checkHistory(iRun.processHistory());
+    // helper class to extract structured data from TTDTC::Frames
+    dataFormats_ = &iSetup.getData(esGetTokenDataFormats_);
+    // helper class to encode layer
+    layerEncoding_ = &iSetup.getData(esGetTokenLayerEncoding_);
+  }
+
+  void ProducerKFin::produce(Event& iEvent, const EventSetup& iSetup) {
+    const DataFormat& dfCot = dataFormats_->format(Variable::cot, Process::sf);
+    const DataFormat& dfZT = dataFormats_->format(Variable::zT, Process::sf);
+    // empty KFin products
+    TTDTC::Streams streamAcceptedStubs(dataFormats_->numStreams(Process::kf) * setup_->numLayers());
+    StreamsTrack streamAcceptedTracks(dataFormats_->numStreams(Process::kf));
+    TTDTC::Streams streamLostStubs(dataFormats_->numStreams(Process::kf) * setup_->numLayers());
+    StreamsTrack streamLostTracks(dataFormats_->numStreams(Process::kf));
+    // read in SFout Product and produce KFin product
+    if (setup_->configurationSupported()) {
+      Handle<TTDTC::Streams> handleStubs;
+      iEvent.getByToken<TTDTC::Streams>(edGetTokenStubs_, handleStubs);
+      const TTDTC::Streams& streams = *handleStubs.product();
+      Handle<vector<TTTrack<Ref_Phase2TrackerDigi_>>> handleTTTracks;
+      iEvent.getByToken<vector<TTTrack<Ref_Phase2TrackerDigi_>>>(edGetTokenTTTracks_, handleTTTracks);
+      const vector<TTTrack<Ref_Phase2TrackerDigi_>>& ttTracks = *handleTTTracks.product();
+      for (int region = 0; region < setup_->numRegions(); region++) {
+        int nStubsSF(0);
+        for (int channel = 0; channel < dataFormats_->numChannel(Process::sf); channel++) {
+          const int index = region * dataFormats_->numChannel(Process::sf) + channel;
+          const TTDTC::Stream& stream = streams[index];
+          nStubsSF += accumulate(stream.begin(), stream.end(), 0, [](int& sum, const TTDTC::Frame& frame){ return sum += frame.first.isNonnull() ? 1 : 0; });
+        }
+        vector<StubSF> stubsSF;
+        stubsSF.reserve(nStubsSF);
+        for (int channel = 0; channel < dataFormats_->numChannel(Process::sf); channel++) {
+          const int index = region * dataFormats_->numChannel(Process::sf) + channel;
+          for (const TTDTC::Frame& frame : streams[index])
+            if (frame.first.isNonnull())
+              stubsSF.emplace_back(frame, dataFormats_);
+        }
+        vector<deque<TTDTC::Frame>> dequesStubs(dataFormats_->numChannel(Process::kf) * setup_->numLayers());
+        vector<deque<FrameTrack>> dequesTracks(dataFormats_->numChannel(Process::kf));
+        int i(0);
+        for (const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack : ttTracks) {
+          if ((int)ttTrack.phiSector() / setup_->numSectorsPhi() != region) {
+            i++;
+            continue;
+          }
+          const int sectorPhi = ttTrack.phiSector() % setup_->numSectorsPhi();
+          deque<FrameTrack>& tracks = dequesTracks[sectorPhi];
```

Comment:
dequesTracks[] is addressed here by sectorPhi, which is 0 or 1. At L147 dequesTracks is dimensioned numChannel(Process::kf), which is kfNumWorker = 2. So this is currently OK. But if anyone changes the cfg parameter kfNumWorker, it will crash. Can this be made safe?

Reply:
Yes, at a certain point I will iterate over this.

Reply:
In meantime, please add comment to NumWorker param in TrackerDTC/python/ProducerES_cfi.py saying "currently only 2 supported".